astro/src/content/blog/building-self-hostable-application.mdx

```diff
@@ -0,0 +1,366 @@
+
+You can also build  automatic downgrading or rolling back of an upgrade. Plan to spend time on the tooling and infrastructure to handle such changes in a self-contained way. 
+
+One workaround for the lack of canarying is offering beta or early access programs to let customers self-select into running versions of your software not yet available to the wider public. Early access installs happen at the discretion of your customer, rather than being in your control, but still can provide valuable feedback. 
+
+The complications around releases mean you need to document and build your systems such that your users have a clear idea of the impact of production upgrades. This is an understood problem and there are several approaches that may apply, including separating the control and data plane functionality and maintaining strict n-1 version backwards compatibility. If you offer rollbacks and depend on a RDBMS, you'll need to make sure schema changes don't destroy or degrade data that would make a rollback impossible, such as dropping a column or changing the column to a data type with less fidelity.
+
+Whatever your release strategy, it must be documented for the developers who build on top of your self-hostable product.
+
+### Release Communication
+
+In addition to the nuts and bolts of how a release happens and what impact it has on your users, to make sure they know about it. After all, if they don't, they won't upgrade.
+
+You must communicate to the user that a new release of your package is available. At FusionAuth, we offer:
+
+* an RSS feed
+* a generic new version email list
+* a security focused email list
+* customer outreach when appropriate
+
+One that we don't offer yet is in-app nagging. Depending on how your users interact with your tool, this may be an option.
+
+Plan to spend some time building out appropriate communication mechanisms.
+
+### Performance
+
+Plan to spend time testing performance across multiple environments before every release. This includes:
+
+* the amount of data in the system
+* hardware variations
+* dependency versions
+* major architectural components you support (such as different databases)
+
+### Understanding Product Usage
+
+Ever used tooling which lets you know what features are used by which users? And which users in a new cohort are similar to users who purchased in the past? Such tooling is foundational for operational efficiencies of SaaS vendors everywhere. Again, that tooling is not for you, unless you build it. The more you report, the higher the impact on your users' privacy. But a high level of privacy is one of the valuable parts of self-hosting.
+
```

Comment:
Could be a good please to use a meme. Mention how users in the wild will have different (generally less) resources than you run and test your software on which can lead to finding more issues and eventually harden your product.
![93ezxw](https://github.com/user-attachments/assets/cb0d162c-fad8-4ffe-becd-646989dd3277)